src/main/java/bio/terra/service/filedata/google/firestore/FireStoreUtils.java

```diff
@@ -29,14 +32,38 @@
             Thread.currentThread().interrupt();
             throw new FileSystemExecutionException(op + " - execution interrupted", ex);
         } catch (ExecutionException ex) {
-            // If this exception has a cause that is a runtime exception, then we rethrow the cause.
-            // Typically, the cause is one of our file system messages we want to see.
-            // If there is no cause or it is not runtime, we wrap the exception in a generic one of ours.
+            // The ExecutionException wraps the underlying exception caught in the Future, so we need
+            // to examine the properties of the cause to understand what to do.
+            // Possible outcomes:
+            // - throw FileSystemRetryException for retryable firestore exceptions to ask the step to retry
+            // - throw FileSystemExecutionException for other firestore exceptions (not retryable)
+            // - rethrow RuntimeExceptions to expose other unexpected exceptions
+            // - throw FileSystemExecutionException to wrap non-Runtime (oddball) exceptions
+
             Throwable throwable = ex.getCause();
+            while (throwable instanceof ExecutionException) {
+                ExecutionException eex = (ExecutionException)throwable;
+                throwable = eex.getCause();
+            }
+            if (throwable instanceof AbortedException) {
+                AbortedException aex = (AbortedException)throwable;
+                // TODO: in general, log + rethrow is bad form. For now, I want to make sure we see these in
+                //  the log as they happen. Once we are comfortable that retry is working properly, we can
+                //  rely on the Stairway debug logging as needed.
+                String msg = "Retrying aborted exception: " + aex.toString();
+                logger.info(msg);
+                throw new FileSystemRetryException(msg);
+            }
+            if (throwable instanceof FirestoreException) {
+                FirestoreException fex = (FirestoreException)throwable;
+                String msg = "Retrying firestore exception: " + fex.toString();
+                logger.info(msg);
```

Comment:
Looking at the code base I see both logging and throwing and just throwing. Usually I prefer not logging at throws since code typically logs at the catch point, so logging at the throw will result in double logging. There are some cases where this isn't true, e.g. code called from a third party library that doesn't log on throw, but those are usually an exceptional case.

In this case, if there's a common catch point that implements the retry (e.g. in `Flight.stepWithRetry()` or in `Step.doStep()`) then you can log there, instead of having to log every place you're throwing and expect something to be retried.

Reply:
Yes, log+throw is generally an anti-pattern. We do it sometimes to make debugging simpler.